tso-server/server/server.go

```diff
@@ -41,46 +49,110 @@ type atomicObject struct {
 
 // TimestampOracle generates a unique timestamp.
 type TimestampOracle struct {
-	ts     atomic.Value
-	ticker *time.Ticker
+	cfg *Config
+
+	ts atomic.Value
 
 	listener net.Listener
 
 	wg       sync.WaitGroup
 	connLock sync.Mutex
 	conns    map[net.Conn]struct{}
+
+	zkElector *zkhelper.ZElector
+	zkConn    zkhelper.Conn
+
+	isLeader int64
+
+	lastSavedTime time.Time
 }
 
-func (tso *TimestampOracle) updateTicker() {
-	// TODO: save latest TS to persistent storage(zookeeper/etcd...)
+func (tso *TimestampOracle) loadTimestamp() error {
+	last, err := loadTimestamp(tso.zkConn, tso.cfg.RootPath)
+	if err == zk.ErrNoNode {
+		// no timestamp node, create later
+		err = nil
+	}
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var now time.Time
+
 	for {
-		select {
-		case <-tso.ticker.C:
-			prev := tso.ts.Load().(*atomicObject)
-			now := time.Now()
-
-			// ms
-			since := now.Sub(prev.physical).Nanoseconds() / 1e6
-			if since > 2*updateTimtestampStep {
-				log.Warnf("clock offset: %v, prev:%v, now %v", since, prev.physical, now)
-			}
-			// Avoid the same physical time stamp
-			if since <= 0 {
-				log.Warn("invalid physical time stamp")
-				continue
-			}
+		now = time.Now()
 
-			current := &atomicObject{
-				physical: now,
-			}
-			tso.ts.Store(current)
+		since := (now.UnixNano() - last) / 1e6
```

Comment:
no need, here we only handle falling behind.  we only check millisecond, if we restart quickly in test, ms may be same, and at same time, if since = 0, we will wait 2 * interval time to guarantee next timestamp is ok.